pkg/cli/cliexecuter.go

```diff
@@ -1,97 +1,168 @@
-	if len(pod.Spec.Containers) == 0 {
-		err := errors.New("create cli executer failed")
-		e.log.Error(err, "No container found in Pod", "Pod Name", pod.Name)
-		return "", "", err
+// AsyncExecute runs the command in a goroutine mapped to the UID
+func (e *Executer) AsyncExecute(uid types.UID, command []string) {
+	e.mapLock.Lock()
+	defer e.mapLock.Unlock()
+	if _, exist := e.routines[uid]; !exist {
+		e.routines[uid] = true
+
+		go func(uid types.UID, command []string) {
+			e.log.Info("fence agent start", "uid", uid, "command", command)
+
+			stdout, stderr, fa_err := e.runner(e.ctx, e.log, uid, command)
+
+			e.log.Info("fence agent done", "uid", uid, "command", command, "stdout", stdout, "stderr", stderr, "err", fa_err)
+
+			// loop to handle only the cases where
+			// - FAR cannot be found, but it does exist
+			// - the status update fails for conflict
+			for {
+				far, err := e.getFARByUID(uid)
+				if err != nil {
+					if !apiErrors.IsNotFound(err) {
+						continue
+					}
+					e.log.Error(err, "could not update status", "FAR uid", uid)
+					break
+				}
+
+				err = e.updateStatus(far, fa_err)
+				if err == nil {
+					e.log.Info("status updated", "FAR uid", uid)
+					break
+				}
+				if apiErrors.IsConflict(err) {
+					e.log.Error(err, "conflict while updating the status", "FAR uid", uid)
+					time.Sleep(1 * time.Second)
+				}
+				e.log.Error(err, "failed to update status", "FAR uid", uid)
+				break
+			}
+		}(uid, command)
 	}
+}
 
-	var (
-		stdoutBuf bytes.Buffer
-		stderrBuf bytes.Buffer
-	)
-
-	//containerName := pod.Spec.Containers[0].Name
-	containerName := "manager"
-
-	// Build the Post request for SPDY (shell) executor
-	req := e.clientSet.CoreV1().RESTClient().Post().
-		Resource("pods").
-		Name(pod.Name).
-		Namespace(pod.Namespace).
-		SubResource("exec").
-		Param("container", containerName)
-
-	req.VersionedParams(&corev1.PodExecOptions{
-		Container: containerName,
-		Command:   command,
-		Stdin:     false,
-		Stdout:    true,
-		Stderr:    true,
-		TTY:       false,
-	}, scheme.ParameterCodec)
-
-	execSPDY, err := remotecommand.NewSPDYExecutor(e.config, "POST", req.URL())
-	if err != nil {
-		e.log.Error(err, "failed building SPDY (shell) executor")
-		return "", "", err
+// Exists checks if there is already a running Fence Agent command mapped to the UID
+func (e *Executer) Exists(uid types.UID) bool {
+	e.mapLock.Lock()
+	defer e.mapLock.Unlock()
+	_, exist := e.routines[uid]
+	return exist
+}
+
+// run runs the command in the container and updates the status of the FAR instance maching the UID
+func run(ctx context.Context, logger logr.Logger, uid types.UID, command []string) (stdout, stderr string, err error) {
+	cmd := exec.CommandContext(ctx, command[0], command[1:]...)
+
+	var outBuilder, errBuilder strings.Builder
+	cmd.Stdout = &outBuilder
+	cmd.Stderr = &errBuilder
+
+	for i := 0; i < 5; i++ {
```

Comment:
Why do we try to run the fence agent only 5 times?
Before this change, we didn't loop on the fence agent output. We used the reconcile back-off to retrigger reconciliation and execute the command. it might be better to continue using that and returning an error in https://github.com/medik8s/fence-agents-remediation/pull/106/files#diff-05147ee3f0d225b5a60754b8b9bff2c569deb0a9d7eaddae33af680291d493a1R198

Reply:
My idea was to consider the failure of the fence agent possible. I agree that the back-off retry is better than a constant delay, but with the previous implementation either the FA works or the reconcile never ends.

We cannot return error in [#106 (files)](https://github.com/medik8s/fence-agents-remediation/pull/106/files#diff-05147ee3f0d225b5a60754b8b9bff2c569deb0a9d7eaddae33af680291d493a1R198). That function only spawn a goroutine and returns immediately.
If we want to treat FA errors the way it was previously, we might let Reconcile loop handle Status Condition "FA failure" to call the fence agent again.

Reply:
>either the FA works or the reconcile never ends.

Correct. If the agent fails then IMHO the reconciliation should not proceed. At least as long it does not block the reconciliation of another far CR.

>We cannot return error in [#106 (files)](https://github.com/medik8s/fence-agents-remediation/pull/106/files#diff-05147ee3f0d225b5a60754b8b9bff2c569deb0a9d7eaddae33af680291d493a1R198). That function only spawn a goroutine and returns immediately.

Yes, I meant the line afterward where ATM we return no error and no requeue.

> If we want to treat FA errors the way it was previously, we might let Reconcile loop handle Status Condition "FA failure" to call the fence agent again.

SGTM but I prefer to limit the dependency on CR status and the reconciliation decisions as much as possible if we can understand whether the agent was succeeded or not.

Reply:
> Yes, I meant the line afterward where ATM we return no error and no requeue.

Why do you want to return an error there? We don't need the Reconcile loop to be re-queued until we know that the FA goroutine has done something (either succeeded or failed).

> SGTM but I prefer to limit the dependency on CR status and the reconciliation decisions as much as possible if we can understand whether the agent was succeeded or not.

I have two in-progress branches, one of those uses `event.GenericEvent` to wake up the controller and check the result, but I preferred this one where the synchronization between main loop and goroutine is triggered by the status update. We could discuss further in the chat what the pros and cons are.